### PR TITLE
Improve pseudo binding for dynamic parameters

### LINKED ISSUE #3984: Dynamic Parameter Doesn't Autocomplete
Consider this function:

```powershell
function Test-DynamicParameter
{
    [CmdletBinding()]
    PARAM
    (
        [string[]]
        $Param1,
        # this parameter seems to cause the problem - if I specify [string] then tab-completion works for -DynamicParameter
        [PSCredential]
        $Param2
    )
    DYNAMICPARAM 
    {
        $Name = "DynamicParameter"
        $ParamAttr = [System.Management.Automation.ParameterAttribute]::new()
        $ParamAttr.Mandatory = $true
        $Attributes = [Collections.ObjectModel.Collection[System.Attribute]]::new()
        $Attributes.Add($ParamAttr)
        $Parameter = [System.Management.Automation.RuntimeDefinedParameter]::new($Name, [string], $Attributes)
        $Dictionary = [System.Management.Automation.RuntimeDefinedParameterDictionary]::new()
        $Dictionary.Add($Name, $Parameter)
        $Dictionary
    }
    PROCESS
    {

        $DynamicParameter
    }
}
```

It seems that auto-completion of the dynamic parameter name is broken by the inclusion of the [PSCredential]-typed parameter. 

This works as expected - hitting tab expands the parameter "-DynamicParameter"
`Test-DynamicParameter -Param1 'a' -<tab>`

This fails - hitting tab does nothing. however, the parameter still works if I type it out
`Test-DynamicParameter -Param1 'a' -Param2 $Credential -<tab>`

If I change the type attribute of Param2 to [string] then tab-completion works for the dynamic parameter

I have observed this happening with other parameter types as well -for instance, the same condition exists if you use [DateTime] for Param2. Also, when using Plaster at the command line to scaffold a new PowerShell project, the parameters defined in the Plaster manifest won't autocomplete. This may also be the same bug.

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      5.1.15063.296
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.15063.296
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes https://github.com/PowerShell/PowerShell/issues/3984
When doing pseudo binding the parameters with the value specified inside the parameter like `-Param:Value`  and non-constant values like variables would be processed as raw text. Now it tries to safely evaluate the values and add the evaluated value to the parameters, allowing binding to work in more scenarios.

It still doesn't work with values that can't be evaluated like `Test-DynamicParameter -Param2 (Get-Credential) -<Tab>`. One could argue that would be by design since the dynamic parameter doesn't have a value but IMO it's still not an ideal user experience.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs [hunk 1]
```diff
@@ -1222,7 +1219,30 @@ private bool PrepareCommandElements(ExecutionContext context)
...
+                                paramToAdd = CommandParameterInternal.CreateParameter(parameter.ParameterName, parameter.Extent.Text);
+                            }
+                            else
+                            {
+                                object value;
+                                if (!SafeExprEvaluator.TrySafeEval(parameter.Argument, context, out value))
+                                {
+                                    value = parameter.Argument.Extent.Text;
+                                }
+                                paramToAdd = CommandParameterInternal.CreateParameterWithArgument(
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs:1236]:
```suggestion
                                if (!SafeExprEvaluator.TrySafeEval(parameter.Argument, context, out value))
                                {
                                    value = parameter.Argument.Extent.Text;
                                }

                                paramToAdd = CommandParameterInternal.CreateParameterWithArgument(
```

File: src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs [hunk 2]
```diff
@@ -1222,7 +1219,30 @@ private bool PrepareCommandElements(ExecutionContext context)
...
+                                    parameterAst: null,
+                                    parameterName: parameter.ParameterName,
+                                    parameterText: parameter.Extent.Text,
+                                    argumentAst: null,
+                                    value: value,
+                                    spaceAfterParameter:false);
+                            }
+                            commandProcessor.AddParameter(paramToAdd);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs:1244]:
```suggestion
                            }

                            commandProcessor.AddParameter(paramToAdd);
```

File: src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs [hunk 3]
```diff
@@ -1248,15 +1268,22 @@ private bool PrepareCommandElements(ExecutionContext context)
                         }
                         else if (_commandElements[commandIndex] is ExpressionAst expression)
                         {
-                            valueToAdd = expression.Extent.Text;
+                            if (!SafeExprEvaluator.TrySafeEval(expression, context, out valueToAdd))
+                            {
+                                valueToAdd = expression.Extent.Text;
+                            }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/PseudoParameterBinder.cs:1274]:
```suggestion
                            if (!SafeExprEvaluator.TrySafeEval(expression, context, out valueToAdd))
                            {
                                valueToAdd = expression.Extent.Text;
                            }

```

**Comment by iSazonov** [comment]
> It still doesn't work with values that can't be evaluated like `Test-DynamicParameter -Param2 (Get-Credential) -<Tab>`.

OutputType()? :-)

**Comment by MartinGC94** [comment]
> > It still doesn't work with values that can't be evaluated like `Test-DynamicParameter -Param2 (Get-Credential) -<Tab>`.
> 
> OutputType()? :-)

Sure, that gets me the type but the problem is that it needs some sort of value otherwise it throws while doing something (I didn't bother looking that much into it).

**Review by iSazonov — APPROVED** [review]

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Pull request merged**